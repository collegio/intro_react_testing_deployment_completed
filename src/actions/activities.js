import uuid from 'uuid';

export const addActivity = ({ name, activity_type = 'running', distance = '0 km' } = {}) => ({
    type: 'ADD_ACTIVITY',
    activity: {
        id: uuid(),
        name,
        activity_type,
        distance
    }
})
  
export const removeActivity = ({ id } = {}) => ({
    type: 'REMOVE_ACTIVITY',
    id
});

export const editActivity = (id, updates) => ({
    type: 'EDIT_ACTIVITY',
    id,
    updates
});