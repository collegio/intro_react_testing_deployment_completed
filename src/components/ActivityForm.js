import React from 'react';

export class ActivityForm extends React.Component {

    constructor(props) {
        super(props);

        this.onNameChange = this.onNameChange.bind(this);
        this.onActivityChange = this.onActivityChange.bind(this);
        this.onDistanceChange = this.onDistanceChange.bind(this);
        this.onSubmitActivity = this.onSubmitActivity.bind(this);

        this.state = {
            name: props.activity ? props.activity.name : '',
            activity_type: props.activity ? props.activity.activity_type : 'walking',
            distance: props.activity ? props.activity.distance : '0 km',
            error: ''
        }
    }

    onNameChange(e) {
        const name = e.target.value;
        this.setState(() => ({ name }));
    };

    onActivityChange(e) {
        const activity_type = e.target.value;
        this.setState(() => ({ activity_type }));
    };

    onDistanceChange(e) {
        const distance = e.target.value;
        this.setState(() => ({ distance }));
    };

    onSubmitActivity(e) {
        e.preventDefault();

        if (!this.state.name) {

            // Error!
            this.setState(() => ({ error: 'Please enter a name first.' }));
        }
        else {

            // Remove the error and Submit!
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                name: this.state.name,
                activity_type: this.state.activity_type,
                distance: this.state.distance
            });
        }
    }

    render() {
        return (
            <div className="activity-form">
                { this.state.error && <p>{this.state.error}</p> }
                <form onSubmit={this.onSubmitActivity}>
                    <label>Name</label><br/>
                    <input type="text" placeholder="Name" value={this.state.name} onChange={this.onNameChange} autoFocus />

                    <label>Sport</label><br/>
                    <select value={this.state.sport_type} onChange={this.onActivityChange}>
                        <option value="running">Running</option>
                        <option value="walking">Walking</option>
                        <option value="cycling">Cycling</option>
                        <option value="swimming">Swimming</option>
                    </select>

                    <label>Distance</label><br/>
                    <input type="text" placeholder="Name" value={this.state.distance} onChange={this.onDistanceChange} autoFocus />

                    <button className="button pull-right">Submit</button>
                </form>
            </div>
        );
    }
}

export default ActivityForm;