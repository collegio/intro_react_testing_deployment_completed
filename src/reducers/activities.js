// the default list of activities, which is empty
const activityDefaultState = [];

export default (state = activityDefaultState, action) => {
    switch (action.type) {

        case 'ADD_ACTIVITY':
            return [
                ...state,
                action.activity
            ];

        case 'REMOVE_ACTIVITY':
            return state.filter(({ id }) => id !== action.id);

        case 'EDIT_ACTIVITY':
            return state.map((activity) => {
                if (activity.id === action.id) {
                    return {
                        ...activity,
                        ...action.updates
                    };
                }
                else {
                    return activity;
                }
            });

        default:
            return state;
    }
};