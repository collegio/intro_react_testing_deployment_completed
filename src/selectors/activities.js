// Get visible activities

export default (activities, { text, activity_type, sort_by }) => {
    return activities.filter((activity) => {

        // match the text search
        const textNameMatch = activity.name.toLowerCase().includes(text.toLowerCase());

        // match the sport type
        let typeMatch = false;
        if (activity_type == 'all' || activity_type == activity.activity_type) {
            typeMatch = true;
        }

        return (textNameMatch && typeMatch);

    }).sort((a, b) => {

        if (sort_by === 'name') {
            return a.name > b.name ? 1 : -1;
        } else if (sort_by === 'distance') {

            // sort in reverse
            return a.distance > b.distance ? -1 : 1;
        }
    });
};