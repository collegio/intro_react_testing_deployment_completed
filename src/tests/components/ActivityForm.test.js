import React from 'react';
import { shallow } from 'enzyme';
import ActivityForm from '../../components/ActivityForm';
import activities from '../fixtures/activities';

test('should render the activity form', () => {
    const wrapper = shallow(<ActivityForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render the activity form with data', () => {
    const wrapper = shallow(<ActivityForm activity={activities[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
    const wrapper = shallow(<ActivityForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set name on input change', () => {
    const value = 'Evening Walk';
    const wrapper = shallow(<ActivityForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('name')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('should set activity type on select change', () => {
    const value = 'cycling';
    const wrapper = shallow(<ActivityForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('select').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('activity_type')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('should set distance on input change', () => {
    const value = '20 km';
    const wrapper = shallow(<ActivityForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('distance')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('should trigger valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ActivityForm activity={activities[0]} onSubmit={onSubmitSpy}/>);

    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });

    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        name: activities[0].name,
        activity_type: activities[0].activity_type,
        distance: activities[0].distance
    });
});