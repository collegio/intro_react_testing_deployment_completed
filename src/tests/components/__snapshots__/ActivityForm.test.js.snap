// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error for invalid form submission 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div
  className="activity-form"
>
  <p>
    Please enter a name first.
  </p>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should render the activity form 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should render the activity form with data 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Evening Walk"
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="2 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set activity type on select change 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set activity type on select change 2`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set distance on input change 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set distance on input change 2`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="20 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set name on input change 1`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should set name on input change 2`] = `
<div
  className="activity-form"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Evening Walk"
    />
    <label>
      Sport
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
    <label>
      Distance
    </label>
    <br />
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="0 km"
    />
    <button
      className="button pull-right"
    >
      Submit
    </button>
  </form>
</div>
`;
